src/chainstate/stacks/db/blocks.rs

```diff
@@ -3662,10 +3661,152 @@ impl StacksChainState {
+        }
+        Ok((cur_blocks, stacking_ops))
+    }
+
+    /// Process any Stacking-related bitcoin operations
+    ///  that haven't been processed in this Stacks fork yet.
+    pub fn process_stacking_ops(
+        clarity_tx: &mut ClarityTx,
+        operations: Vec<StackStxOp>,
+        processed_total: u64,
+    ) -> Result<Vec<StacksTransactionReceipt>, Error> {
+        let mut all_receipts = vec![];
+        let mut cost_so_far = clarity_tx.cost_so_far();
+        for stack_stx_op in operations.into_iter() {
+            let StackStxOp {
+                sender,
+                reward_addr,
+                stacked_ustx,
+                num_cycles,
+                block_height,
+                txid,
+                burn_header_hash,
+                ..
+            } = stack_stx_op;
+            let result = clarity_tx.connection().as_transaction(|tx| {
+                tx.run_contract_call(
+                    &sender.into(),
+                    &QualifiedContractIdentifier::boot_contract("pox"),
+                    "stack-stx",
+                    &[
+                        Value::UInt(stacked_ustx),
+                        reward_addr.as_clarity_tuple().into(),
+                        Value::UInt(u128::from(block_height)),
+                        Value::UInt(u128::from(num_cycles)),
+                    ],
+                    |_, _| false,
+                )
+            });
+            match result {
+                Ok((value, _, events)) => {
+                    if let Value::Response(ref resp) = value {
+                        if !resp.committed {
+                            debug!("StackStx burn op rejected by PoX contract.";
+                                   "txid" => %txid,
+                                   "burn_block" => %burn_header_hash,
+                                   "contract_call_ecode" => %resp.data);
+                        }
+                        let mut execution_cost = clarity_tx.cost_so_far();
+                        execution_cost
+                            .sub(&cost_so_far)
+                            .expect("BUG: cost declined between executions");
+                        cost_so_far = clarity_tx.cost_so_far();
+
+                        let receipt = StacksTransactionReceipt {
+                            transaction: TransactionOrigin::Burn(txid),
+                            events,
+                            result: value,
+                            post_condition_aborted: false,
+                            stx_burned: 0,
+                            contract_analysis: None,
+                            execution_cost,
+                        };
+
+                        all_receipts.push(receipt);
+                    } else {
+                        unreachable!(
+                            "BUG: Non-response value returned by Stacking STX burnchain op"
+                        )
+                    }
+                }
+                Err(e) => {
```

Comment:
Are there cases where this is a non-panicking error?

Reply:
Yeah -- if the block cost budget is exceeded.

Reply:
Yes -- this makes deep forks more expensive, because they must apply the stacking operations that occur between the parent burn block and the current burn block. _Not_ counting that towards the block's compute budget is much more consequential for the network (instead of possibly making certain deep forks impossible, it is a huge DDOS vector). There's things we could do to limit this -- checking the account balance "for free" before invoking the contract and placing a high threshold on BTC stacking ops could place both a theoretical and practical limit on the number of operations that would need to be processed.

Reply:
> There's things we could do to limit this -- checking the account balance "for free" before invoking the contract and placing a high threshold on BTC stacking ops could place both a theoretical and practical limit on the number of operations that would need to be processed.

Yeah -- I think we'll want to impose some rules on the BTC-based Stacking operations to bound the total number of Stacking operations that will be considered, no matter how deep the Stacks fork is.  In particular, there's two cost concerns here:

* Bounding the number of MARF reads required to load up the history of Stacking operations (this can be achieved by setting `earliest-possible-burn-block` appropriately).
* Bounding the execution time of all Stacking operations such that they, collectively, do not exceed a single Stacks block's budget.

I think there's a pretty wide array of strategies we can implement to achieve the latter.  Some ideas come to mind:

* Prioritize the top-N Stacking operations by BTC tx fee, if there are too many to fit into a Stacks block (i.e. if there are more than N of them, where N is a derived protocol constant equal to the maximal number of calls to `stack-stx` that can fit into a block).
* Prioritize the top-N Stacking operations by amount Stacked
* Prioritize the top-N Stacking operations by STX transaction fee paid
* Take the last N Stacking operations that occur in the Bitcoin chain
* Sample N Stacking operations without replacement, using the VRF seed of `burnchain-tip` as the randomness seed
* Require that all on-Bitcoin Stacking operations stack for only 1 reward cycle, assuming that a Bitcoin block will always contain less than or equal to N Stacking transactions

Reply:
Okay, I see you addressed this below by making it so that only the penultimate burnchain block's stacking operations will be applied.